lib/distributed_mutex.rb

```diff
@@ -19,6 +19,7 @@ def initialize(key, redis: nil, validity: DEFAULT_VALIDITY)
     @using_global_redis = true if !redis
     @redis = redis || Discourse.redis
     @mutex = Mutex.new
+    @lock_monitor = Monitor.new
```

Comment:
This is a bit tricky because `@lock_monitor` must be passed around together with `@redis` to ensure all threads using the `@redis` object have the same `@lock_monitor` instance too.